Parse WAL file into WAL records

sh-soltanpour opened the pull request:
This PR is addressing issues #296, #299, #300, #301.

It introduces the infrastructure for parsing a WAL (Write-Ahead Log) file into WAL record data structures. It supports multiple PostgreSQL versions, from 13 to 17, by utilizing official magic values.

Key Features:
- Parses each page header, record header, and the associated data.
- Handles WAL records that span multiple pages.
- Implements a factory pattern (wrapper method) to manage version-specific structural changes across different PostgreSQL versions.

Benefits: Merging this PR will enable the parsing of WAL files into a list of WAL records. This parsed data can be leveraged in the future for rewriting WAL files or extracting the logical meaning of the records.

jesperpedersen commented:
Lets get this squashed and force pushed

jesperpedersen commented:
https://github.com/pgmoneta/pgmoneta/blob/main/doc/DEVELOPERS.md#multiple-commits

sh-soltanpour commented:
done

jesperpedersen commented:
CI

jesperpedersen commented:
Start by cleaning up the `"` `#include` it should be `<>`

jesperpedersen commented:
You should do a `doc/WAL.md` document that describe how everything works.

Put it in the developer guide as well

jesperpedersen commented:
See https://github.com/pgmoneta/pgmoneta/blob/main/doc/DEVELOPERS.md how to generate the manuals, and API documentation.

We need everything documented

jesperpedersen commented:
And, make sure that constants are aligned, e.g.

```
#define MY_KEY      0x01
#define MY_LONG_KEY 0x02
```

The same for struct comments

jesperpedersen commented:
Try and keep the struct comments on the same line as well, like

```
struct server
{
   char name[MISC_LENGTH];             /**< The name of the server */
   char host[MISC_LENGTH];             /**< The host name of the server */
   int port;                           /**< The port of the server */
```

jesperpedersen commented on doc/WAL.md:
```diff
@@ -0,0 +1,39 @@
+
+# WAL Reader
+
+This document provides an overview of the `wal_reader` tool, particularly focusing on the `parse_wal_file` function, which is the main entry point for parsing Write-Ahead Log (WAL) files. Write now this function only parses the given WAL file and prints the description of each record but in future, it will be integrated with the other parts of the code.
```
Write now this -> Right now, this

jesperpedersen commented:
Did you install the documentation requirements ? You should see the warnings during the API documentation generation.

Remember to add WAL.md to the developer guide.

And, fix indentation and the CI error

sh-soltanpour commented:
> Did you install the documentation requirements ? You should see the warnings during the API documentation generation.
> 
> Remember to add WAL.md to the developer guide.
> 
> And, fix indentation and the CI error

Ok yeah I didn't have it installed, after installing it started showing me the warnings. Working on fixing them

jesperpedersen commented:
I havn't looked at the strlcpy issue - but there is pgmoneta_append()

sh-soltanpour commented:
Yeah, I tried to fix it by including `string.h` but it didn't work. Will replace it.

jesperpedersen commented:
Remember to squash and rebase

jesperpedersen commented on doc/WAL.md:
```diff
@@ -0,0 +1,39 @@
+# WAL Reader
+
+This document provides an overview of the `wal_reader` tool, particularly focusing on the `parse_wal_file` function, which is the main entry point for parsing Write-Ahead Log (WAL) files. Right now this function only parses the given WAL file and prints the description of each record but in future, it will be integrated with the other parts of the code.
+
```
Can you explain the different WAL format from PostgreSQL 13 to PostgreSQL 17 ?

Is there an easy place to see that ?

sh-soltanpour replied:
The only difference I could notice was in the `pgmoneta_wal_is_bkp_image_compressed` function. Just added the explanation to this part.

jesperpedersen replied:
Well, I mean, the different PostgreSQL have different fields depending on the version...

* PostgreSQL 13: 0xD106
* PostgreSQL 14: 0xD10D
* PostgreSQL 15: 0xD110
* PostgreSQL 16: 0xD113
* PostgreSQL 17: 0xD116 (currently)

You can see commits like https://github.com/postgres/postgres/commit/2b5819e2b4d6c69676c61ee799fd9590be2308ce that will bump the WAL identifier

jesperpedersen replied:
So, I would like to see a description of how those different formats are supported through the https://github.com/pgmoneta/pgmoneta/blob/main/src/include/pgmoneta.h#L226 field

sh-soltanpour replied:
Okay, I see. The magic values you mentioned are currently stored in the `xlp_magic` in the `xlog_page_header_data` struct, but I don't do any processing based on them and they're just stored in the struct.
The code was able to parse the sample WAL files I generated for testing on Postgresql 13 -> 17, but I would also double-check the commits you sent to note any other fields that may not be covered in my tests

jesperpedersen replied:
As you can see `wal_level` was added to `struct xl_end_of_recovery` and `struct CheckPoint` - that is available in 0xD116, but not in earlier versions

jesperpedersen replied:
So, you should explain how you handle the parsing and the different version since the structs will have different fields

sh-soltanpour replied:
I'm thinking about using a wrapper struct with a factory pattern. We can implement it for `xl_end_of_recovery` and use the same approach anytime a change is made in one of the structs. What do you think about this?
```
struct xl_end_of_recovery_v16 {
    timestamp_tz end_time;
    timeline_id this_timeline_id;
    timeline_id prev_timeline_id;
};

struct xl_end_of_recovery_v17 {
    timestamp_tz end_time;
    timeline_id this_timeline_id;
    timeline_id prev_timeline_id;
    int wal_level;
};

struct xl_end_of_recovery {
    int pg_version;
    union {
        struct xl_end_of_recovery_v16 v16;
        struct xl_end_of_recovery_v17 v17;
    } data;
    void (*parse)(struct xl_end_of_recovery* wrapper, const void* rec);
    char* (*format)(struct xl_end_of_recovery* wrapper, char* buf);
};
xl_end_of_recovery* create_xl_end_of_recovery(int pg_version) {
    xl_end_of_recovery* wrapper = malloc(sizeof(xl_end_of_recovery));
    wrapper->pg_version = pg_version;
    
    if (pg_version >= 17) {
        wrapper->parse = parse_v17;
        wrapper->format = format_v17;
    } else {
        wrapper->parse = parse_v16;
        wrapper->format = format_v16;
    }
    
    return wrapper;
}

void parse_v16(xl_end_of_recovery* wrapper, const void* rec) {
    memcpy(&wrapper->data.v16, rec, sizeof(struct xl_end_of_recovery_v16));
}

void parse_v17(xl_end_of_recovery* wrapper, const void* rec) {
    memcpy(&wrapper->data.v17, rec, sizeof(struct xl_end_of_recovery_v17));
}

char* format_v16(xl_end_of_recovery* wrapper, char* buf) {
    struct xl_end_of_recovery_v16* xlrec = &wrapper->data.v16;
    return pgmoneta_format_and_append(buf, "tli %u; prev tli %u; time %s",
                                      xlrec->this_timeline_id, xlrec->prev_timeline_id,
                                      pgmoneta_wal_timestamptz_to_str(xlrec->end_time));
}

char* format_v17(xl_end_of_recovery* wrapper, char* buf) {
    struct xl_end_of_recovery_v17* xlrec = &wrapper->data.v17;
    return pgmoneta_format_and_append(buf, "tli %u; prev tli %u; time %s; wal_level %d",
                                      xlrec->this_timeline_id, xlrec->prev_timeline_id,
                                      pgmoneta_wal_timestamptz_to_str(xlrec->end_time),
                                      xlrec->wal_level);
}
```

jesperpedersen replied:
Yeah, that could be a solution.

Do you have a list of changes between the different versions ? Do we need a lot of these parse/format functions ?

jesperpedersen replied:
Unfortunately, we have to go through the history of `src/include/access/xlog_internal.h` and investigate the commits that bumped the magic number.

There is no official list of changes :disappointed:

sh-soltanpour replied:
Yeah I guess that would take a lot of time to check all structs manually :(

jesperpedersen replied:
Yeah, it isn't ideal.

But, commits like https://github.com/postgres/postgres/commit/f83d709760d84253b539ee454ce9ce1ceb9ad9ac changed existing `struct`s and added new ones

jesperpedersen replied:
Just to make it clear -  we only need to support the official magic numbers.

So, you need to find those commits, and then diff between them in one go

jesperpedersen replied:
E.g. 0xD106 -> 0xD10D, 0xD10D -> 0xD110, and so on

sh-soltanpour replied:
Ok I see, so in total there should be like 4-5 commits that should be diffed, this is not hard and can be handled

jesperpedersen replied:
You can document the actual commit ids for each magic number change in the developer guide if you want.

Some of the commits are just year bumps, and so on

jesperpedersen replied:
E.g.

```
0xD116: 2b5819e2b4d6c69676c61ee799fd9590be2308ce
...
```

sh-soltanpour replied:
This is the list of commits for official changes.
I'm looking at the differences between version 17 and 16 and it's huge! ~4000 files changed with ~3500 of them in the src directory and ~700 in the `src/include`, so manual checking of every change in every file is impossible. I'm going to only check file changes in `src/include/access` files that are ~90 file changes for the last commit. Does the strategy sound good to you? or maybe you have better suggestions to narrow it down?
```
PostgreSQL 13: 0xD106: c6b92041d38512a4176ed76ad06f713d2e6c01a8
PostgreSQL 14: 0xD10D: 08aa89b326261b669648df97d4f2a6edba22d26a
PostgreSQL 15: 0xD110: 8b1dccd37c71ed2ff016294d8f9053a32b02b19e
PostgreSQL 16: 0xD113: 6af1793954e8c5e753af83c3edb37ed3267dd179
PostgreSQL 17: 0xD116: 402b586d0a9caae9412d25fcf1b91dae45375833
```

jesperpedersen replied:
Yes, it should be ok to just verify the `src/include/access/` directory, so

```
git diff 6af1793954e8c5e753af83c3edb37ed3267dd179..402b586d0a9caae9412d25fcf1b91dae45375833 src/include/access/
```

jesperpedersen replied:
.. and you can use a lot of `| grep -v`s to get it down even further

sh-soltanpour replied:
I have gone through the commits version by version and compiled a list of changes here: https://elastic-traffic-761.notion.site/Postgres-WAL-diff-5d7f873c9d204ff79cb6779262c76326.

I tried my best not to miss any changes that are related to Pgmoneta, but since the process an overlook is possible. Can you please review it and let me know if it's good to go? I can start implementing the solution above for the documented structs.

jesperpedersen replied:
Yeah, it looks good - you can put this information into the developer guide

jesperpedersen commented on src/include/wal/relpath.h:
```diff
@@ -0,0 +1,56 @@
+#define DEFAULTTABLESPACE_OID 1663
+#define GLOBALTABLESPACE_OID  1664
+
+#define PG_MAJORVERSION                "14" // TODO: Update with pgmoneta server version
```
Is only PostgreSQL 14 supported ?

sh-soltanpour replied:
This is just used for generating the description string

jesperpedersen replied:
As long as we support 13 -> 17

jesperpedersen commented on src/include/wal/relpath.h:
```diff
@@ -0,0 +1,56 @@
+#define GLOBALTABLESPACE_OID  1664
+
+#define PG_MAJORVERSION                "14" // TODO: Update with pgmoneta server version
+#define CATALOG_VERSION_NO             "202107181"
```
This should probably be tied to the version number

sh-soltanpour replied:
yeah

jesperpedersen commented on src/include/wal/relpath.h:
```diff
@@ -0,0 +1,56 @@
+#define RELPATHPERM(rnode, forknum) \
+        RELPATHBACKEND(rnode, INVALID_BACKEND_ID, forknum)
+
+char*
```
Missing documentation

jesperpedersen commented on src/include/wal/rm_btree.h:
```diff
@@ -0,0 +1,345 @@
+ * @param record The decoded XLog record.
+ * @return A pointer to the description in buf.
+ */
+char* btree_desc(char* buf, struct decoded_xlog_record* record);
```
Formatting

jesperpedersen commented on src/include/wal/rm_btree.h:
```diff
@@ -0,0 +1,345 @@
+ * @param info The operation info byte.
+ * @return A string identifying the operation.
+ */
+const char* btree_identify(uint8_t info);
```
const + formatting

jesperpedersen commented on src/include/wal/rm_generic.h:
```diff
@@ -0,0 +1,60 @@
+ * @param record A pointer to the decoded WAL record.
+ * @return A pointer to the buffer containing the description.
+ */
+char* generic_desc(char* buf, struct decoded_xlog_record* record);
```
Formatting

jesperpedersen commented:
You can look at the API documentation - there are some `@brief` statements missing

jesperpedersen commented:
Once you are happy then take it out of "Draft" mode

jesperpedersen commented:
Lets prefix all functions in the .h files with `pgmoneta_wal_`

sh-soltanpour commented:
> You can look at the API documentation - there are some `@brief` statements missing

Do we want `@brief` for all functions and structs? or just functions?

jesperpedersen commented:
Ideally, everything -- you can look at the API documentation, and see what is missing.

I havn't found a doxygen check that will fail a build if there is documentation missing, so we have to do best effort manually - not ideal but ...

jesperpedersen commented:
Remember to update your forum thread with your progress

jesperpedersen commented:
See CI... ParseAbortRecordV14 -> parse_abort_record_v14

Are we ready for non-draft once CI is green ?

sh-soltanpour commented:
Yeah sure, Im on a short vacation for a few days, will take care of CI and
developer docs this week, and then yeah it would be ready for non-draft
then.

On Fri, Sep 6, 2024 at 4:11 PM Jesper Pedersen ***@***.***>
wrote:

> See CI... ParseAbortRecordV14 -> parse_abort_record_v14
>
> Are we ready for non-draft once CI is green ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pgmoneta/pgmoneta/pull/355#issuecomment-2334742080>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF5BS7CNWXZUAJYC3D4YBC3ZVID7DAVCNFSM6AAAAABMJKZNTCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMZUG42DEMBYGA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

jesperpedersen commented:
These are some general comments.

* You should rewrite the commit message
* Move .h files to src/include/wal/walfile/ -- implementation can stay in src/libpgmoneta/wal/

* Its "return type \newline function definition"
* Its typedef, prototypes, \#define, structs and then functions -- if it is too hard just leave it be
* There is a `struct ... {`
* There are `enum ... {`

You can see if you play around with uncrustify's configuration if it can help you.

For \#include it is pgmoneta first (sorted) then a newline then system ones (sorted)

There is missing function documentation in `rm_hash.h` and `rm_xact.h`

Align documentation comments within \#define groups, and for struct vaiables.

Replace `printf` with `pgmoneta_log_trace` for now -- we circle back to this once we have the debugging features in place.

Otherwise, than the minor stuff above, this is a great start for this new area of pgmoneta !

sh-soltanpour commented:
> These are some general comments.
> 
> * You should rewrite the commit message
> * Move .h files to src/include/wal/walfile/ -- implementation can stay in src/libpgmoneta/wal/
> * Its "return type \newline function definition"
> * Its typedef, prototypes, #define, structs and then functions -- if it is too hard just leave it be
> * There is a `struct ... {`
> * There are `enum ... {`
> 
> You can see if you play around with uncrustify's configuration if it can help you.
> 
> For #include it is pgmoneta first (sorted) then a newline then system ones (sorted)
> 
> There is missing function documentation in `rm_hash.h` and `rm_xact.h`
> 
> Align documentation comments within #define groups, and for struct vaiables.
> 
> Replace `printf` with `pgmoneta_log_trace` for now -- we circle back to this once we have the debugging features in place.
> 
> Otherwise, than the minor stuff above, this is a great start for this new area of pgmoneta !

All these comments are fixed now. For some of them that needed huge formatting and I couldn't find anything in uncrustify, I used ChatGPT and revised the response manually, there's a chance that some of them are overseen during my review but most of them are fixed now.

jesperpedersen commented:
Try and search for `struct XYZ {` and `union {` -- I don't know why uncrustify doesn't pick that up... -- they should be on next line...

Rebease too

sh-soltanpour commented:
Ahh I misunderstood your message and changed everything to `struct ... {` format, will fix it

sh-soltanpour commented:
Just to double check, we want version this:

struct ABC {
...
}
enum ABC {
....
}


or this?
struct ABC
{
...
}
enum ABC
{
....
}

sh-soltanpour commented:
> Try and search for `struct XYZ {` and `union {` -- I don't know why uncrustify doesn't pick that up... -- they should be on next line...
> 
> Rebease too

Fixed and rebased.

jesperpedersen commented:
Merged.

Thanks for your contribution !
